base/compiler/typelimits.jl

```diff
@@ -268,8 +268,11 @@ function type_more_complex(@nospecialize(t), @nospecialize(c), sources::SimpleVe
     return true
 end
 
+union_count_abstract(x::Union) = union_count_abstract(x.a) + union_count_abstract(x.b)
+union_count_abstract(@nospecialize(x)) = !isdispatchelem(x)
+
 function issimpleenoughtype(@nospecialize t)
-    return unionlen(t) <= MAX_TYPEUNION_LENGTH && unioncomplexity(t) <= MAX_TYPEUNION_COMPLEXITY
+    return unionlen(t)+union_count_abstract(t) <= MAX_TYPEUNION_LENGTH && unioncomplexity(t) <= MAX_TYPEUNION_COMPLEXITY
```

Comment:
I was thinking this would get included in the complexity score computation. Just wondering reasoning for this way, though either seems good.

Reply:
Yes, I think that would make more sense, but it doesn't increase the union complexity over the maximum. The current complexity of the type is 2; if we're willing to move that to 4 then it can work.